man/man1/tabpp.1
.TH TABPP 1NEMO "15 May 1998"
.SH NAME
tabpp \- Poynter-Pickett spectral line catalogue manipulator
.SH SYNOPSIS
\fBtabpp\fP \fBspecies=\fPs1,s2... \fBmode=\fP [parameter=value]
.SH DESCRIPTION
\fItabpp\fP manipulates the Poynter-Pickett catalogue of 
Submillimeter, Millimeter, and Microwave spectral lines
in the frequency range between 0 and 10000 Ghz (i.e.
wavelengths longer then 30 um).
The catalogue can be used as a planning guide or as an
aid in the identification and analysis of observed
spectral lines. The information listed for each spectral 
line includes the frequency and its estimated error, the
intensity, lower state energy, and quantum number assignment.
.PP
The catalogue has been constructed by using 
theoretical least squares fits of published spectral
lines to accepted molecular models. The associated
predictions and their estimated errors are based upon
the resultant fitted parameters and their covariances. Future versions
of this catalogue will add more atoms and molecules and 
update the present listing (331 species) as new data appear.
.SH PARAMETERS
The following parameters are recognized in any order if the keyword is also
given:
.TP 20
\fBspecies=\fP
Species to check for. The default is \fBall\fP. To get a listing of
all valid (known) species in the catalogue, try the shorthand
command "\fB tabpp all species name | more \fP".
Species examples: \fIH-atom, D-atom, OH, C-13-O, 
HNO3, HCOOD, O2, HOCl, NNH+\fP. Note names are case sensitive.
.TP
\fBmode=\fP
Output mode. Either the \fBspecies\fP can be listed, or the 
spectral \fPlines\fP. Minimum match applies. No default.
.TP
\fBcol=\fP
Columns to print out. In \fBlines\fP mode (see previous) you can choose
any of \fIfreq, err, lgint, dr, elow, gup, tag, qnform, qn1, qn2\fP,
or choose \fIall\fP.
In \fBspecies\fP mode any of \fI tag, name, nline, qlog, version\fP, or
\fIall\fP. In \fBlines\fP mode, you can also print out any of the
\fBspecies\fP  columns.
\fI ** Note **\fP Currently selection by column names is not active. In
\fBlines\fP mode  you can
either choose \fIall\fP, or something else, in which case a shorter 
version with  \fIfreq, err, lgint, dr, elow, gup, tag\fP,
will be output, with the \fIname\fP and computed \fImu2\fP in the last 
two columns.
.TP
\fBfreq=\fP
Frequency range to select. Two numbers are required, with an optional
third character string specifying the frequency
units given and requested (the catalogue's numbers are listed in
Mhz, and this is the default). Note that if the units of \fIfreq\fP 
are changed, so will be units of \fIerr\fP and \fPlgint\fP. 
We only support \fBMHz, GHz\fP and \fPTHz\fP (case insensitive).
See example below. 
.TP
\fBcutoff=\fP
Lower line intensity cutoff applied to the output. Any LGINT's smaller than
the cutoff, will not be printed. Default: all will be printed.
.TP
\fBselect=\fP
A comma separated list of numerical comparisons expressions that are
logically AND'ed. Each expression much be of the form \fBa OPER b\fP,
where \fBOPER\fP is any of "<", "<=", ">", ">=", "==", "!=", and
\fBa\fP and \fBb\fP are  \fInemoinp(3NEMO)\fP 
expressions that may also contain references to the column names
if they are preceded by a "%" symbol. 
Example: 
.nf
            \fIselect="sqrt(%Rad) < %lum**2/(%D_0 + %R_0)"\fP
.fi
.TP
\fBfmt=\fP
\fI**not implemented**\fP
.TP
\fBtab=\fP
Table name, in which output is gathered. Default: terminal.
\fINot implemented\fP
.TP
\fBdir=\fP
Up to three string can be given, separated with comma's.
The first one is the name of the unix directory where the catalogue
is located. If none is provided, the environment variable
\fB$PP_DIR\fP is tried, or the local directory, if all
else fails.
The second string is the (C) format specification how the
species filesname are constructed (default is
\fBc%s.cat.Z\fP, where \fB%s\fP is the 6 digit species
identification TAG). The third string is the filename of
the species-directory filename. It defaults to \fBcatdir.cat\fP.
It is not advised to use this keyword, as \fItabpp\fP should have
been setup properly.
.PP
Note: the old pre-1994 catalog will need \fBc%s.ed4.gz\fP and \fBcatdir.ed4\fP.
The maryland version needs TAB_PP set to /lma/spectra.
.SH EXAMPLES
The following command lists all species present in the catalogue
.nf
    % tabpp species=all mode=species col=name | more
or shorthand, since these are the first three program keywords:
    % tabpp all species name | more
.fi
this can be followed up by a frequency search of a few species:
.nf
    % tabpp species=HCOOH,HOCL mode=lines col=mu2 freq=80,115,ghz | more
  or shorthand:
    % tabpp HCOOH,HOCL lines name,freq,err 80,115,ghz | more
.fi
.SH FORMAT
The line catalogue files are in the Footran format:
.nf
.ta +1i
 FREQ,  ERR,  LGINT,DR, ELOW,  GUP,TAG,QNFORM, QN1, QN2
(F13.4, F8.4, F8.4, I2, F10.4, I3, I7, I4,     6I2, 6I2, 1X)

FREQ	frequency of line (by default in Mhz, but you can change it)
ERR	estimated or experimental error in FREQ
LGINT	base 10 logarithm of the integrated intensity 
DR	degree of freedom in the rotatational partition function
ELO	lower state energy relative to ground state
GUP	upper state degeneracy
TAG	species tag or molecular identifier. Negative means laboratory freq.
QNFMT	identifies format of the quantum numbers in QN
QN'	quantum numbers of the upper state coded according to ANFMT
QN''	quantum numbers of the lower state 
.nf
.PP
The species-directory file is in the Footran format:
.nf
TAG,    NAME, NLINE, QLOG,  VERSION
(I6,1X, A14,  I5,    7F7.4, I2)
.fi
.PP
A special column name, \fBmu2\fP identifies the electric dipole matrix,
which can be printed out too. It's units are Debye**2
.PP
.nf
.ta +1i
TAG	species tag number in format \fBWWWWXXXX\fP, \fBWWW\fP=weight, \fBXXX\fP=ordinal
NAME	ASCII name for the species
NLINE	number of lines for this species
QLOG	log10 of the partition function for 300, 225, 150, 75, 37.5, 18.75, 9.375 K
VERSION	version number of this species
.fi
.SH UNITS
.nf
.ta +1i +1i
freq	MHz     	(optionally GHz, THz; see \fBfreq=\fP keyword)
err	MHz      	(optionally GHz, THz)
lgint	nm^2.Mhz  	(optionally nm^2.GHz, nm^2.THz)
elow	cm^-1
qlog	...
mu2	Debye^2         (computed from other quantities)
.fi
.SH TIMING
The full catalogue of 206 species contains 20Mb of compressed ascii data, 
and hence browsing through the whole catalogue will take time. For example,
to find ALL lines between 80 and 81 Ghz, we could issue the following
commands:
.nf
    % time tabpp all lines mu2 80,81,Ghz > log1
    2092.020u 118.850s 40:33.46 90.8% 0+116k 2300+29io 2248pf+0w
.fi
whereas using the (slow) \fBselect=\fP keyword took much longer:
.nf
    % time tabpp all lines mu2 select="%freq < 81000,%freq > 80000" > log2

.fi
.SH SEE ALSO
awk(1), grep(1v), tablovas(1NEMO)
.PP
JPL Molecular Spectroscopy: http://spec.jpl.nasa.gov/
.PP
\fIThe Fitting and Prediction of Vibration-Rotation Spectra 
with Spin Interactions\fP,
H. M. Pickett. J. Molec. Spectroscopy 148, 371-377 (1991). 
.PP
\fISubmillimeter, Millimeter, and Microwave Spectral Line Catalogue\fP,
R.L. Poynter and H.M. Pickett. 
Applied Optics 24, 2335-2240 (1985).
This catalogue is also available from these authors as a magnetic
tape recorded in card images and as a set of microfishe records,
or from anonymous ftp \fIftp://spec.jpl.nasa.gov:/pub/catalog/\fP.
.PP
Molecular collision rates for line radiative transfer: 
http://www.strw.leidenuniv.nl/~moldata/
.PP
Anderson, Herbst and De Lucia's list for methanol lines (unknown ref)
.PP
Atomic Data for Astrophysics: http://www.pa.uky.edu/~verner/atom.html
.PP
Table of Isotopes: http://isotopes.lbl.gov/isotopes/toi.html
.PP
Chemical MIME draft: http://www.ch.ic.ac.uk/chemime2.html
.PP
Infrared Line Lists: http://www.mpe.mpg.de/www_ir/ISO/linelists/
.PP
http://www.splatalogue.net/
.PP
http://www.astro.uni-koeln.de/cdms/
.PP
https://astroquery.readthedocs.io/en/latest/atomic/atomic.html
.PP
https://astroquery.readthedocs.io/en/latest/nist/nist.html
.SH ENVIRONMENT
.nf
.ta +1i
PP_DIR             	Home directory of the catalogue
.fi
.SH FILES
.nf
.ta +1.5i
$PP_DIR           	home directory of the catalogue (e.g. $MIRCAT/jplcat or $NEMODAT/jplcat)
catdir.cat        	species data for the new (>1994) edition
cWWWXXX.cat.gz       	XXXth (gzip compressed) line data for a species with mol.weight WWW
catdir.ed4       	species data for the old 4th edition
cWWWXXX.ed4.Z    	XXXth (compressed) line data for a species with mol.weight WWW
src/scripts/get_jplcatalog	script to install the catalog via SMA
.fi
.SH BUGS
Cannot search for isotopes yet - this complicates the search, since 
HCN may appear as HCN, HCN-15 in the PP catalogue. The Lovas
catalogue seems to use a different isotope notation, e.g.
HC15N, where it can also be confused with multiplicities, e.g.
HC9N (here they mean 9 C's on a row).
.SH AUTHOR
Peter Teuben
.SH HISTORY
.nf +1i +4.5i
29-feb-92	V1.0 Created - data has 206 species         	PJT
24-mar-92	V1.1 compute mu^2 to appease lgm/joe 	PJT
3-oct-92	V1.2 implemented select= 	PJT
18-jan-95	V1.3 made it work for the new 1994 edition of PP_DIR=/lyra/pp
11-sep-95	V1.3a new unofficial edition, now 228 species
	22 new species: SiC, SiC-v1, SiC-13, HCP, DCP, CH3OCH3, HCCNC,
	  HCCNC-v7, HCCNC-v6, CCS, C-13CCS, CC-13S, CCS-34, 
	  CH3OCHO-A, CH3OCHO-E, OS-34-O, HC6, HOBr-79, HOBr-81, HC7N, HC9N, HC11N
	corrected spelling error H2C-13-S
9-oct-95	V1.3b updated, now 272 species / 1041209 lines
15-may-98	V1.4 merged two diverged versions, now 331 species / 1747809 lines
17-aug-05	[miriad $MIRCAT/jplcat resurrected] 338 species / 2209176 lines	PJT
17-jul-11	V1.5 fixed strcpy for more modern gcc compilers	PJT
22-sep-14       default for $NEMODAT/jplcat 
.fi


src/kernel/tab/tabpp.c
/*
 * TABPP: Poynter-Pickett spectral line catalogue browser
 *          See the man page for many more details.
 *
 *      29-feb-92  V1.0     A toy version created                  PJT
 *	 9-mar-92      a   bit more doc, allow Thz units
 *			   handle $ENVNAME in dir=		   PJT
 *	12-mar-92      c   added cutoff=			   PJT
 *      20-mar-92      d   warning when not using mhz - to appease joe
 *      24-mar-92  V1.1    added computation of \mu, the electric
 *                         dipole matrix element                   PJT
 *	 3-oct-92  V1.2    implemented select=                     PJT
 *	19-jan-95  V1.3    fixed up for the new (>1994) catalog    PJT
 *				(rely on zcat to do .Z and .gz)
 * 	11-sep-95  V1.3a   fixed parsing error $PP_DIR		   pjt
 *	 9-oct-95      b   new .cat extension, instead of .ed4	   pjt
 *      15-may-98  V1.4    merged two confusing version            pjt
 *      17-jul-11  V1.5    fixed copying into colval[]             pjt
 */
/**************** NEMO * UNIX INCLUDE FILES **********************************/ 

#include <stdinc.h>
#include <getparam.h>
#include <filefn.h>
#include <strlib.h>
#include <cgs.h>		/* CGS constants */

/**************** COMMAND LINE PARAMETERS *************************************/

string defv[] = {
    "species=\n                    List of species [all]",
    "mode=???\n                    Mode of output {species|lines}",
    "col=\n                        List of columns to see [all]\n\
        In species mode: (tag,name,nline,qlog,version)|all\n\
        In lines mode:   (freq,err,lgint,dr,elow,gup,tag,qnform,qn1,qn2)|all",
    "freq=\n    Ranges in frequency to show, and units ([Mhz]|Ghz|THz)",
    "cutoff=\n  Cutoff in LGINT (log of intensity)",
    "tab=\n     Output table [terminal](** not implemented **)",
    "select=\n  C-style selection criteria (** not implemented **)",
    "fmt=\n     Alternative format for column output (** not implemented **)",
    "dir=$PP_DIR,c%s.cat,catdir.cat\n  Dir, species-fmt, sp-dirfile",
    "VERSION=1.5\n               17-jul-11 PJT",
    NULL,
};

string usage = "Poynter-Pickett spectral line catalogue browser";

/**************** GLOBAL VARIABLES ********************************************/

real freqmin=0.0,
     freqmax=999999999.999,
     factor=1.0,
     lfactor=0.0,
     cutoff=-999999.9;

string fmt1 = "     FREQ         ERR     LGINT  DR     ELOW   GUP   TAG   NAME   MU^2\n";
string fmt2 = "%14.4f %8.4f %8.4f %2d %10.4f %3d %7d %s %f\n";
string fmt3 = "__________________________________________________________\n";
string fmt4 = "   NAME         TAG    Nlines\n";
string fmt5 = 
"   TAG NAME,         NLINE QLOG_1 QLOG_2 QLOG_3 QLOG_4 QLOG_5 QLOG_6 QLOG_7 VERSION\n";
string fmt6 =
"        FREQ     ERR    LGINT DR    ELOW   GUP   TAG QNFORM QN1(6)   QN2(6)\n";

string fmt7 = "%14.4f%8.4f%8.4f%s\n";

#define MAXSPECIES 350

struct species {
    char tag[7];
    char name[15];
    int nline;
    float qlog[7];
    int version;
} species[MAXSPECIES];

string colnam[] = {
    "freq", "err", "lgint", "dr", "elow", "gup", "tag", "qnform", "mu2",
    NULL};

string colval[9] ;
int scount, lcount;

string defdir = "/lma/spectra";
string deffmt = "c%s.ed4";
string defcat = "catdir.ed4";

extern string *burststring(string, string);
extern bool scanopt(string, string);

local string *burstcommand(string);
local char *mysubstr(string,int *, int);
local void strlower(char *s);

/****************************** START OF PROGRAM *****************************/

nemo_main()
{
    stream instr;
    string mode, tab;
    string *dirs, *usespecies, *cols, *freqs, *sels, names;
    real tmp;
    int  i;
    bool lmode, Qcolall, Qselall;
    char sname[128], sfile[128], *cp;
    string dname;

    tab = getparam("tab");
    mode = getparam("mode");
    lmode = (*mode=='l' || *mode=='L');

    names = getparam("species");
    usespecies = burststring(names,", ");
    cols = burststring(getparam("col"),", ");
    Qcolall = (xstrlen(cols,sizeof(string))<=1 || streq(cols[0],"all"));
    dirs = burststring(getparam("dir"),", ");
    sels = burststring(getparam("select"),",");   /* note only comma separated, not space */
    Qselall = (xstrlen(sels,sizeof(string))<=1);
    if(xstrlen(dirs,sizeof(string)) != 4)
        error("dir=$PP_DIR,sp-fmt,catname");
    if(hasvalue("cutoff")) cutoff = getdparam("cutoff");

    cp = dirs[0];
    if (*cp == '$') {
        cp++;
        if (*cp)
             dname = getenv(cp);
        else
             dname = getenv("PP_DIR");  /* If name given: try PP_DIR */
        if (dname==NULL) dname = sconc(".","/");    /* or else, use current */
    } else {
    	dname = dirs[0];
    }
    if (dname[strlen(dname)-1] != '/')	/* UNIX: make sure '/' is last char */
        dname = sconc(dname,"/");	/* for later catenation */

    freqs = burststring(getparam("freq"),", ");
    switch(xstrlen(freqs,sizeof(string))-1) {
     default:
        warning("Bad parsing: assume whole freq range...\n");
     case 0:
        break;
     case 1:
        nemoinpr(freqs[0],&freqmin ,1);
        break;
     case 2:
        nemoinpr(freqs[0],&freqmin ,1);
        nemoinpr(freqs[1],&freqmax ,1);
        break;
     case 3:
        nemoinpr(freqs[0],&freqmin ,1);
        nemoinpr(freqs[1],&freqmax ,1);
        strlower(freqs[2]);
        if(streq(freqs[2],"mhz"))
            factor = 1.0;
        else if (streq(freqs[2],"ghz")) {
            warning("Units in GHz: note freq, err and lgint are modified");
            factor = 0.001;
        } else if (streq(freqs[2],"thz")) {
            warning("Units in GHz: note freq, err and lgint are modified");
	    factor = 0.000001;
        } else
            error("%s is an illegal frequency unit: must be (Mhz,Ghz,Thz)",
                  freqs[2]);
        break;
    }
    lfactor = log10(factor);
    if(freqmin>freqmax) {
          warning("Toggling freq's since they must be sorted");
          freqmin = tmp;
          freqmin = freqmax;
          freqmax = tmp;
    }

    read_species(sconc(dname,dirs[2]), cols, lmode);


    if(lmode) {         /* mode=line ? */
        if (check_species(usespecies) == 0)
            error("No species left to search for");
        if(*names==NULL || streq(names,"all"))
            dprintf(1,"[Be patient, this operation may take a while]\n");
	if (Qcolall) {
            dprintf(0,fmt6);
	} else
	    dprintf(0,fmt1);
        for (i=0; i<scount; i++) {      /* loop over all species */
            sprintf(sname,dirs[1],species[i].tag);
            sprintf(sfile,"%s%s",dname,sname);
            if(*names && !streq(names,"all")) {
                if(!scanopt(names,species[i].name))
                    continue;
            }
            read_lines(sfile,&species[i],Qcolall,sels);
        }
    }
}

patchzero(char *name)
{
    while (*name) if(*name == ' ') *name++ = '0'; else name++;
}

patchnull(char *name)
{
    while (*name) if(*name == ' ') *name++ = '\0'; else name++;
}



/**********************************************************************/
read_species(fname, cols, lmode, Qall)		/* read data from input file */
string fname;
string *cols[];
bool lmode, Qall;
{
    char line[128], tag[6+1], name[14+1], note[1+1];
    int  nline, version, tag1, tag2;
    float qlog[7];
    stream instr;
    int i, n;

    instr = stropen(fname,"r");
    dprintf (2,"Reading speciesfile \n");
    if(!lmode) {
        if (Qall)
            dprintf(0,fmt5);
        else
            dprintf(0,fmt4);
    }

    scount = 0;
    while (fgets(line,128,instr) != NULL) {
        line[strlen(line)-1] = '\0';
        n = 0;
        strcpy(tag,mysubstr(line,&n,6));  tag[6] = '\0';
        patchzero(tag);  /* replace blanks by '0' */
        n++;
        strcpy(name,mysubstr(line,&n,14));  name[14] = '\0';
        patchnull(name);   /* replace blanks by '\0' */
        nline = atoi(mysubstr(line,&n,5));
        for(i=0; i<7; i++)
            qlog[i] = atof(mysubstr(line,&n,7));
        version = atoi(mysubstr(line,&n,2)); 
        strcpy(note,mysubstr(line,&n,1)); note[1] = '\0';

        strcpy(species[scount].tag,tag);
        strcpy(species[scount].name,name);
        species[scount].nline = nline;
        for(i=0; i<7; i++)
                    species[scount].qlog[i] = qlog[i];
        species[scount].version = version;

        scount++;
        lcount += nline;

        if (!lmode) {
            if (Qall)
                printf("%s\n",line);
            else
                printf("%-14s %-6s %5d\n",name,tag,nline);
        }
    }
    dprintf(0,"%s: %d species read, with a total of %d lines\n",
              fname,scount,lcount);
    strclose(instr);
}

read_lines(fname,spp,Qall,sels)	/* read species data from input file */
string fname;			/* pointed to by 'instr'     */
struct species *spp;            /* the species this applies to */
bool Qall;                      /* print all of one line? */
string *sels;                   /* list of selection criteria applied to each line */
{
    char line[128], zcatcmd[256], *cp;
    string ext;
    float freq, err, lgint, elow, eup, mu2;
    int   dr, gup, tag, qnform;
    int   qn1[6], qn2[6];
    int i, n;
    bool pipe;
    stream instr;

    dprintf (2,"Reading lines file %s\n",fname);
    ext = extension(fname);
    if (streq(ext,"Z") || streq(ext,"gz")) {
        pipe = TRUE;
        sprintf(zcatcmd,"zcat %s",fname);
        instr = popen(zcatcmd,"r");
    } else {
        pipe = FALSE;
        instr=stropen(fname,"r");
    }
    for (i=0; i<9; i++)  colval[i] = allocate(32);

    while (fgets(line,128,instr) != NULL) {          /* read row from line catalogue */
        line[strlen(line)-1] = '\0';
	dprintf(3,"%s\n",line);
        n = 0;
        /***  "freq", "err", "lgint", "dr", "elow", "gup", "tag", "qnform",   ***/
	/*    1041.4215   .0026 -8.7541 3   69.4458 23  46005 30311 3 8      11 3 9  */
        /*    1420.4058  0.0000 -9.0612 0    0.0000  3  -1001  22 1 1         1 0    */
	/*1234567890123........xxxxxxxx--xxxxxxxxxx---xxxxxxx----                    */

        cp = mysubstr(line,&n,13); freq =    atof(cp);  strcpy(colval[0],cp);
        cp = mysubstr(line,&n,8);  err  =    atof(cp);  strcpy(colval[1],cp);
        cp = mysubstr(line,&n,8);  lgint=    atof(cp);  strcpy(colval[2],cp);
        cp = mysubstr(line,&n,2);  dr   =    atoi(cp);  strcpy(colval[3],cp);
        cp = mysubstr(line,&n,10); elow =    atof(cp);  strcpy(colval[4],cp);
        cp = mysubstr(line,&n,3);  gup  =    atoi(cp);  strcpy(colval[5],cp);
        cp = mysubstr(line,&n,7);  tag  =    atoi(cp);  strcpy(colval[6],cp);
        cp = mysubstr(line,&n,4);  qnform  = atoi(cp);  strcpy(colval[7],cp);
        for(i=0; i<6; i++)
            qn1[i]  =  atoi(mysubstr(line,&n,2));
        for(i=0; i<6; i++)
            qn2[i]  =  atoi(mysubstr(line,&n,2));

	/* compute \mu when still in Mhz units */
        eup = elow + freq/29978.3;      /* E_upper in cm^-1 */
	mu2 = 1.502e11 * pow(10.0,(double)lgint+spp->qlog[0])
            * exp(eup/208.55) / (freq*freq*gup);
        sprintf(colval[8],"%g",mu2);

	/* change some units for output */
        freq *= factor;
        err  *= factor;
        lgint += lfactor;

        if(freqmin <= freq && freq <= freqmax && lgint>cutoff) {
            if(num_select(colnam, colval, sels))
                if(Qall)
                    printf(fmt7,freq,err,lgint,&line[29]);
                else
                    printf(fmt2,freq,err,lgint,dr,elow,gup,tag,spp->name,mu2);
        }
        dprintf(1,"%s",line);
    }
    if(pipe) pclose(instr);
    else   strclose(instr);
}

check_species(string *names)
{
    int i, n = xstrlen(names,sizeof(string))-1;
    bool found;
    string *sn;

    if (n==0) return(1);
    if (streq(names[0],"all")) return(1);

    for (sn=names; *sn; sn++) {
        found = FALSE;
        for(i=0; i<scount; i++) {
            if(streq(*sn,species[i].name)) {
                found = TRUE;
                break;
            }
        }
        if(!found) {
            warning("Species %s not present in catalogue\n",*sn);
            n--;
        }
    }
    return(n);
}



local char *mysubstr(char *line,int *n, int l)
{
    local char word[100];

    strncpy(word,&line[*n],l);
    word[l] = 0;
    *n += l;

    return word;
}

local string *burstcommand(string cmd)
{
    stream pstr;
    string *bc;
    char line[100];
    int n=0;

    bc = (string *) allocate((MAXSPECIES+1)*sizeof(string *));

    dprintf(3,"BURSTCOMMAND: CMD=%s\n",cmd);
    pstr = popen(cmd,"r");
    while (fgets(line,100,pstr) != NULL) {
        line[strlen(line)-1] = '\0';
        dprintf(3,"BURSTCOMMAND> %s\n",line);
        bc[n] = sconc(line,"");
        n++;
        if(n==MAXSPECIES) {
            warning("Too many species; MAXSPECIES=%d\n",MAXSPECIES);
            break;
        }
    }
    pclose(pstr);
    bc[n] = NULL;
    return(bc);
}

/* STRLOWER: convert a string to lower case */

void strlower(char *s)
{
    while (*s) {
       if(isupper(*s)) *s = tolower(*s);
       s++;
    }
}
       

